design/continuous-delivery.md

```diff
@@ -0,0 +1,578 @@
+
+**Asset Source Staging**
+
+During synthesis, the CDK app will _copy the sources_ of all assets from their original location on disk into the cloud assembly output directory. This allows the cloud assembly to be self-contained and is important for the CI/CD use cases (both internal and external) where the cloud assembly is the only artifact needed after build is complete.
+
+**Asset Providers**
+
+Users should be able to vend custom asset providers to allow customizing how assets are being referenced or packaged.
+
+For example, a company might have an internal system that manages software artifacts. They can internally vend custom implementations for the `lambda.Code` and `ecs.ContainerImage` classes which will allow users to reference these artifacts and synthesize placeholders into the cloud assembly, which will later be resolved during the publishing stage and identified through a user-defined unique identifier.
+
+**Environment-agnostic Stacks**
+
+When a stack is defined, users can specify `env` (account and/or region).
+
+If `account` and/or `region` use the pseudo references `Aws.ACCOUNT_ID` and `Aws.REGION`, respectively, the stack is called "environment-agnostic". Certain features in the CDK, like VPC lookups for example, are not supported for environment-agnostic stacks since the specific account/region is required during synthesis.
+
+The proposal described in [PR#4131](https://github.com/aws/aws-cdk/pull/4131) suggests that environment-agnostics stacks cannot be deployed using the CDK. However, it also proposes that the default behavior for `env` will be to use ("inherit") the CLI configured environment when a stack is deployed through `cdk deploy`.
+
+This means that the only way to produce environment-agnostic templates will be to explicitly indicate it when a stack is defined.
+
+Since the specific account and region are required when resolving asset consumption and publishing locations, the current plan is for the default asset store implementation to **fail if assets are used from environment-agnostic stacks**. Again, bear in mind that the current behavior (where the default is environment-agnostic stacks) is going to be changed.
+
+## Bootstrapping
+
+The CDK already has a dedicated tool for bootstrapping environments called **`cdk bootstrap`**. An environment is bootstrapped once, and from that point, it is possible to deploy CDK apps into this environment.
+
+Environment bootstrapping doesn't have to be performed by the development team, and does not require deep knowledge of the application structure, besides the set of accounts and regions into which the app needs to be deployed.
+
+The current implementation only provisions an S3 bucket, but in order to be able to continuously deploy CDK stacks that use asses, we will need the following resources:
+
+For publishing:
+
+* **S3 Bucket**: for file asset and CloudFormation templates
+* **ECR Repository**: for docker image assets
```

Comment:
I just want to clarify one thing:

There can be many Docker assets as part of the same CDK app. Are we saying all those different images that are the results of building those Docker assets will be stored in one ECR repository?

Reply:
Yes.